gym_xiangqi/pieces.py

```diff
@@ -0,0 +1,180 @@
+    BLACK = 1
+    DEAD = 0
+    ALIVE = 1
+
+    # Must set this attributes in subclasses
+    color = 0
+    row = 0
+    col = 0
+    state = 0
+    image = None
+
+    def move(self, newRow, newCol):
+        """
+        Take one move among given piece's allowed moves
+        Update piece's coordinates internally
+        """
+        raise NotImplementedError
+
+
+class King(Piece):
+    """
+    This piece is equivalent to King.
+    It is called "Jiang" or "Shuai" meaning Governor (red) and General (black) respectively.
+    Only one piece exists in each side
+    Can only move 1 unit of space orthogonally within the special square area
+    """
+
+    def __init__(self, color, row, col):
+        self.color = color
+        self.row = row
+        self.col = col
+        self.state = Piece.ALIVE
+        # TODO: add image for pygame rendering
+        #  self.image =
+
+    def move(self, newRow, newCol):
+        self.row = newRow
+        self.col = newCol
+
+
+class Queen(Piece):
+    """
+    This piece is not quite equivalent to Queen but for the sake of simplicity we will regard as so.
+    It is called "Shi" meaning counselor/scholar.
+    2 pieces exist in each side
+    Can only move 1 unit of space diagonally within the special square area
+    """
+
+    def __init__(self, color, row, col):
+        self.color = color
+        self.row = row
+        self.col = col
+        self.state = Piece.ALIVE
+        # TODO: add image for pygame rendering
+        #  self.image =
+
+    def move(self, newRow, newCol):
+        self.row = newRow
+        self.col = newCol
+
+
+class Bishop(Piece):
+    """
+    This piece is similar to Bishop.
+    It is called "Shiang" meaning minister (red) and elephant (black)
+    2 pieces exist in each side
+    This piece cannot cross the river
+    Moves 2 unit of space diagonally
+    """
+
+    def __init__(self, color, row, col):
+        self.color = color
+        self.row = row
+        self.col = col
+        # TODO: add image for pygame rendering
+        #  self.image =
+
+    def move(self, newRow, newCol):
+        self.row = newRow
+        self.col = newCol
+
+
+class Knight(Piece):
+    """
+    This piece is almost identical to knight.
+    It is called "Ma" meaning horse.
+    2 Pieces in each side.
+    Moves just like knights in chess. (The "L" shape move)
+    The difference is that it cannot jump over pieces
+    """
+
+    def __init__(self, color, row, col):
+        self.color = color
+        self.row = row
+        self.col = col
+        # TODO: add image for pygame rendering
+        #  self.image =
+
+    def move(self, newRow, newCol):
+        self.row = newRow
+        self.col = newCol
+
+
+class Rook(Piece):
+    """
+    This piece is identical to Rook.
+    It is called "Chuh" meaning chariot
+    2 pieces in each side.
+    Moves just like a Rook in chess; As many as you want horizontally and vertically.
+    """
+
+    def __init__(self, color, row, col):
```

Comment:
Hmm do you actually need to repeat the initialization method? Can we just do super init?

Reply:
We can probably do the same with the move method

Reply:
Fixed 👍